### ISSUE TITLE
bug: `keycloak-authz` plugin without `resource_registration_endpoint` when `lazy_load_paths` is true returns 500 error 

### ISSUE THREAD

**Issue by kayx23** [apache/apisix#10708]
### Current Behavior

_Issue separated out from [this report](https://github.com/apache/apisix/issues/10682#issuecomment-1867094155)._

Configure keycloak and configure a route as such (`lazy_load_paths` :`true`):

```shell
$ curl http://127.0.0.1:9180/apisix/admin/routes/1 -X PUT -d '
{
    "uri": "/*",
    "plugins": {
        "authz-keycloak": {
            "lazy_load_paths": true,
            "discovery": "http://192.168.1.83:8080/realms/quickstart-realm/.well-known/openid-configuration",
            "client_id": "your-client-name",
            "client_secret": "your-client-secret"
        }
    },
    "upstream": {
        "type": "roundrobin",
        "nodes": {
            "httpbin.org": 1
        }
    }
}'
```

Request an access token from keycloak and send a request to the route with the access token:

![image](https://github.com/apache/apisix/assets/39619599/9b1558d5-6fec-4124-971f-8c5614c8b762)

because `resource_registration_endpoint` is neither explicitly configured nor presented in the discovery doc:

```text
2023/12/22 00:16:00 [warn] 49#49: *15931 [lua] plugin.lua:1153: run_plugin(): authz-keycloak exits with http status code 503, client: 172.24.0.1, server: _, request: "GET /anything/test HTTP/1.1", host: "127.0.0.1:9080"
2023/12/22 00:16:38 [error] 50#50: *17428 [lua] authz-keycloak.lua:505: authz_keycloak_resolve_resource(): Unable to determine registration endpoint., client: 172.24.0.1, server: _, request: "GET /anything/test HTTP/1.1", host: "127.0.0.1:9080"
2023/12/22 00:16:38 [error] 50#50: *17428 lua entry thread aborted: runtime error: /usr/local/apisix/apisix/plugins/authz-keycloak.lua:580: attempt to get length of local 'permission' (a number value)
stack traceback:
coroutine 0:
172.24.0.1 - - [22/Dec/2023:00:16:38 +0000] 127.0.0.1:9080 "GET /anything/test HTTP/1.1" 500 249 0.129 "-" "curl/7.64.1" - - - "http://127.0.0.1:9080"
/usr/local/apisix/apisix/plugins/authz-keycloak.lua: in function 'evaluate_permissions'
/usr/local/apisix/apisix/plugins/authz-keycloak.lua:778: in function 'phase_func'
/usr/local/apisix/apisix/plugin.lua:1148: in function 'run_plugin'
/usr/local/apisix/apisix/init.lua:725: in function 'http_access_phase'
```

The 503 is expected: 

![image](https://github.com/apache/apisix/assets/39619599/716f95dc-9ba7-4c52-b0c4-45494b77c00e)

But the 500?

### Expected Behavior

Not 500.

### Environment

- APISIX version (run `apisix version`): 3.7.0

**Comment by shreemaan-abhishek** [comment]
@luoluoyuyu would you like to take a look when you are free?

**Comment by lakshya8066** [comment]
@shreemaan-abhishek I can look into this

**Comment by luoluoyuyu** [comment]
Hi @lakshya8066 
Have you tried to reproduce the error?

**Comment by luoluoyuyu** [comment]
Hi @kayx23 
I tried it and it returned 503
![image](https://github.com/apache/apisix/assets/82068138/a683c7ce-6cdb-4ae0-8ed1-f6d9a4f12696)

**Comment by lakshya8066** [comment]
Yes I am trying to reproduce the issue. Having errors 403 error while sending the request. I think it is because of some issue with the JWT token I fetched.
@kayx23 Is the `ACCESS_TOKEN` fetched from http://192.168.1.83/realms/quickstart-realm/protocol/openid-connect/token

**Comment by kayx23** [comment]
@lakshya8066 Hi, yes.

**Comment by kayx23** [comment]
But not sure if 403 is due to invalid JWT. I thought it'd be 401 unauthorized? Could the 403 come from APISIX?

**Comment by lakshya8066** [comment]
@kayx23 Yes the error might be coming from APISIX. 

This is the route config I used for apisix and I have a keycloack instance running on port 8080
```
curl http://127.0.0.1:9180/apisix/admin/routes/1 -X PUT -d '
{
    "uri": "/*",
    "plugins": {
        "authz-keycloak": {
            "token_endpoint": "http://localhost:8080/realms/lak-realm/protocol/openid-connect/token",
            "client_id": "lak-client"
        }
    },
    "upstream": {
        "type": "roundrobin",
        "nodes": {
            "httpbin.org": 1
        }
    }
}'
```
I used this command get the access token: 
`curl --location 'http://localhost:8080/realms/lak-realm/protocol/openid-connect/token' \
--header 'Content-Type: application/x-www-form-urlencoded' \
--data-urlencode 'client_id=lak-client' \
--data-urlencode 'client_secret=Tqq0WmryhDxUMDSGo7KFx3bSt8c13jQ5' \
--data-urlencode 'grant_type=password' \
--data-urlencode 'username=lak' \
--data-urlencode 'password=pass123'`

but while making a curl request like: `curl --location 'http://127.0.0.1:9080/ip'
--header 'Authorization: Bearer <ACCESS_TOKEN>'` I get this error: `{"error":"access_denied","error_description":"not_authorized"}`

This looks like it is coming from this section
![image](https://github.com/apache/apisix/assets/73181332/e3763108-01dd-466b-9c17-919ddcee069b)
But I have not set policy_enforcement_mode="ENFORCING" in the config so it should not matter if I give permissions parameter as an input or not.

**Comment by luoluoyuyu** [comment]
@lakshya8066 
The default value is `ENFORCING`
![image](https://github.com/apache/apisix/assets/82068138/96a5748a-1a4a-45de-aa3f-ec2afc772b2d)

**Comment by kayx23** [comment]
My guess: `lazy_load_paths` is false by default and requires permission to be set. Mentioned here: https://apisix.apache.org/docs/apisix/plugins/authz-keycloak/#permissions

Why not just configure the options I configured in the PR description?

**Comment by lakshya8066** [comment]
You are right @luoluoyuyu @kayx23 
I got it to work and get past through apisix but now the kyecloack instance running in docker is unable to reach the discovery URL. I'll debug that and get back

**Comment by kayx23** [comment]
If you by chance started APISIX using the quickstart script from getting started, the IP address needs to be your 192.x.x.x inet address and not loopback.

**Comment by lakshya8066** [comment]
> If you by chance started APISIX using the quickstart script from getting started, the IP address needs to be your 192.x.x.x inet address and not loopback.

@kayx23 Yes, you were correct. It worked after running apisix after building from source.

**Comment by lakshya8066** [comment]
I found the issue here, In the function `authz_keycloak_resolve_resource` function, we return a 503 error code when a resource_registration_endpoint is not found. But this is not being caught anywhere in the function `evaluate_permissions`. Instead, we are checking is `permission` is `nil` or not and if it is, then pass a 503 error code.

![image](https://github.com/apache/apisix/assets/73181332/1ea336a3-6513-4d00-a8ef-9f7e2e8af08f)

So I think returning `nil` here instead of 503 is the solution. I have tested it as well.

![image](https://github.com/apache/apisix/assets/73181332/031ebdf4-019c-4a58-90dd-689b1e4a10f0)
